File: workspace/java/algorithms/datastructure/src/main/java/host/fairy/recursion/Factorial.java
/*****************************************************
 * @software: IntelliJ IDEA
 * @author: Lionel Johnson
 * @contact: https://fairy.host
 * @organization: https://github.com/FairylandFuture
 * @datetime: 2025-09-06 14:16:02 UTC+08:00
 ****************************************************/
package host.fairy.recursion;

/**
 * 计算阶乘
 *
 * @author Lionel Johnson
 */
public class Factorial {
    public static void main(String[] args) {
        Integer result = factorial(5);
        System.out.println("result = " + result);
    }
    
    /**
     * 计算 n 的阶乘
     *
     * @param n 非负整数
     * @return n的阶乘
     * @throws IllegalArgumentException 当 n 为负数时抛出异常
     */
    public static Integer factorial(Integer n) throws IllegalArgumentException {
        if (n < 0) {
            throw new IllegalArgumentException("n must be non-negative");
        }
        
        if (n == 0 || n == 1) {
            return 1;
        }
        
        return n * factorial(n - 1);
    }
}

File: workspace/python/playground/playground-data-structure-algorithm/playground-01-data-structure/recursion/factorial.py
# coding: UTF-8
"""
@software: PyCharm
@author: Lionel Johnson
@contact: https://fairy.host
@organization: https://github.com/FairylandFuture
@datetime: 2025-09-06 14:22:57 UTC+08:00
"""


def factorial(n: int) -> int:
    """
    计算n的阶乘
    :param n: 非负整数
    :type n: int
    :return: n的阶乘
    :rtype: int
    """
    if n < 0:
        raise ValueError("n must be a non-negative integer")
    if n == 0 or n == 1:
        return 1
    return n * factorial(n - 1)


if __name__ == "__main__":
    print(factorial(5))
